source/s3_auto_ranged_put.c

```diff
@@ -119,10 +119,15 @@ static int s_s3_auto_ranged_put_pause(
     struct aws_s3_meta_request *meta_request,
     struct aws_s3_meta_request_resume_token **resume_token);
 
-static bool s_process_part_info(const struct aws_s3_part_info *info, void *user_data) {
+static bool s_process_part_info_synced(const struct aws_s3_part_info *info, void *user_data) {
     struct aws_s3_auto_ranged_put *auto_ranged_put = user_data;
+    struct aws_s3_meta_request *meta_request = &auto_ranged_put->base;
 
-    struct aws_string *etag = aws_strip_quotes(auto_ranged_put->base.allocator, info->e_tag);
+    ASSERT_SYNCED_DATA_LOCK_HELD(auto_ranged_put);
+
+    struct aws_s3_put_part_info *part = aws_mem_calloc(meta_request->allocator, 1, sizeof(struct aws_s3_put_part_info));
```

Comment:
something to consider: part list is already by value, so instead of allocating and then setting data into array, we can ensure list size, get a ptr to element at given part index and update its fields

Reply:
Remember how Dengke needed to change the byte-buffers to be heap-allocated instead of by-value because the list resizes, but we pass pointers around?: https://github.com/awslabs/aws-c-s3/pull/318

Anyway, I didn't change the weird way we store checksum, it still needs to be on the heap. But I figured I'd just put this whole struct on the heap (with the byte-buffer by-value within it). I took advantage of the on-the-heap nature in another place too: see [line 883](https://github.com/awslabs/aws-c-s3/pull/322/files#diff-1b77484ee8493ab0ab8653a326f4bb0ac42a569e63261469b653ab544d27d99aR883) of the s_skip_parts_from_stream_loop()